Custom fan processing

lcy0x1 opened the pull request:
# Contents

Restructure fan processing classes and code
Add custom fan recipe type
Add data-driven fan processing types
Add data-driven entity conversions using fans
Add example datapack in test/resources to demonstrate how it works

# TODO
1. Currently I use empty bucket for custom fan recipe category icon. Question mark could be more symbolic
2. Blocks in custom fan recipe JEI does not have query interactions

# Classes
1. `InWorldProcessing.Type` -> `AbstractFanProcessingType`
2. `Type.SPLASH` etc -> `TypeSplash` etc in `content.contraptions.processing.fan`
4. EntityTransformHelper - perform the entity conversion logic and particle/sound effects
5. CustomFanNetworkManager - load and sync custom fan processing types and custom entity conversions

# Logic changes
1. Now all fan processing recipes except custom processing use the same RECIPE_WRAPPER
2. `morphType` logic is now in `FanProcessingType` as well. I added several methods in `AirCurrentPartcle` to help with it.
3. Entity conversion was moved from `affectEntity` in `HAUNTING` to `tickAffectedEntity` in `AirCurrent`
4. Add recipe reference in `renderAttachedBlock` in JEI's `ProcessingViaFanCategory`

lcy0x1 committed:
extract fan processing types

lcy0x1 committed:
add custom fan processing recipe type

lcy0x1 committed:
add config structure for custom processing type

lcy0x1 committed:
add JEI support for custom fan processing types

lcy0x1 committed:
add network sync for custom fan processing types

lcy0x1 committed:
use encodeStart()

lcy0x1 committed:
add test datapack

lcy0x1 committed:
add entity transform

lcy0x1 committed:
debug loading

lcy0x1 committed:
debug particle

lcy0x1 committed:
switch to resource location for IDs

lcy0x1 committed:
Merge branch 'mc1.18/dev' into custom_fan_processing

lcy0x1 committed:
fix method order

lcy0x1 committed:
fix import order

lcy0x1 committed:
revert import changes for CreateJEI

lcy0x1 committed:
apply editor config formatting

lcy0x1 committed:
apply editor config formatting

lcy0x1 committed:
follow java naming conventions

lcy0x1 committed:
use objects instead of resource locations when possible

lcy0x1 committed:
add comments explaining how the config works

lcy0x1 committed:
add custom recipe lang entry

lcy0x1 committed:
color parse check

lcy0x1 committed:
throw error when block not found

PepperCode1 commented:
This is better, but I see that fan types are still identified by `String`s and not `ResourceLocation`s.

lcy0x1 commented:
This is because I want to be backward compatible. The names are used by the NBTs as well, which the original fan types are named as enums, such as `SPLASHING`, `HAUNTING`, etc.

If I'm going to change it, should I just leave the original ones as String, and use ResourceLocation for the new ones?

PepperCode1 commented:
Which exact lines in the code require backward compatibility?

lcy0x1 commented:
The 5 original fan types use string for id (`NONE`, `BLASTING` `SMOKING` `SPLASHING` `HAUNTING`). Even if I require ResourceLocation for ID for custom fan types, I still need to store them as String. Do you want me to use ResourceLocation anyway? I can do that, but I would still store the old ones as String, otherwise the old saves will corrupt. Those names are used for NBT of ItemEntity as well.

PepperCode1 commented:
You did not provide the exact lines in the code, so I will have to make a guess at how to ensure backwards compatibility while using `ResourceLocation`s: when the NBT is deserialized and an old type name is discovered (one that does not contain a colon, I assume), convert it to the new name by lowercasing it and prepending `Create.ID + ":" + `.

lcy0x1 commented:
I will do that. I wasn’t sure if I should also change the tags. But then it would be incompatible with PR #2035. Should I also take that into consideration, or create another branch to deal with that problem?

PepperCode1 commented:
Any merge conflicts can be resolved later as we do not yet know if that pull request will be merged in its current form or not.

lcy0x1 commented:
I think I fixed it. However, I also need to modify advancements (as they also use the names). Not sure how that affects player saves.

Also I removed tags in block predicates, because I find that tags might not be loaded before fan types.

lcy0x1 committed:
use resource locations for ID

lcy0x1 committed:
add blocks and fluids to ingredients

lcy0x1 commented:
I was going to add JEI support for block types as well, but I see that fluids are bugged in Create JEI. I will do it after that's fixed

lcy0x1 committed:
add block item display

lcy0x1 commented:
It turns out that JEI is bugged in 9.3.2.92 and I switched to 9.2.1.69 to test it.

Added the block in front of fan as an item displayed at the upper right corner of the block

if fluid is used, both the fluid and its bucket can be used to refer to the recipe, but the recipe only render the bucket form

PepperCode1 commented:
There are three things left for you to do:
- Merge the latest commits from upstream.
- Make it so the name of a custom fan type is retrieved from the location of the file that defines that type instead of a field inside the file.
- Fix the remaining formatting changes (some arguments still use snake case).

There are many other changes I would like to see, but since they are either small, scattered, or specific, I will do them myself. Once I am finished, I will send another comment here and we can discuss if this pull request is ready to be merged.

lcy0x1 committed:
Merge branch 'mc1.18/dev' into custom_fan_processing

lcy0x1 commented:
- [x] merge upstream
- [x]  use file name as fan type name
- [x] fix formatting irregularities
- [x] test

lcy0x1 committed:
regenerate resources

lcy0x1 committed:
fix import and naming format

lcy0x1 committed:
use file name as name

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
reduce white space diff

lcy0x1 committed:
Merge branch 'mc1.18/dev' into custom_fan_processing

lcy0x1 committed:
solve conflict in JEI

lcy0x1 committed:
solve white space diff in JEI

lcy0x1 committed:
solve white space diff in JEI

lcy0x1 committed:
solve white space diff in JEI

lcy0x1 committed:
Merge branch 'mc1.18/dev' into custom_fan_processing

gr8pefish commented:
If it helps, know that both myself and many others would be stoked to have this available!
If I know it'll be merged I'd even potentially offer to update it to have no conflicts myself.

lcy0x1 commented:
I would love to have this as well, but I don’t have to deliver it anymore. Could you help me out with this?

gr8pefish commented:
If @PepperCode1 is willing to merge it, I can do that! I just want to make sure the work would be worthwhile is all :D

Attack8 commented:
What is this meant to accomplish? It looks cool but I’m not sure I understand it fully!

Attack8 commented:
I have merged the mc1.18/dev branch with this pr in #4748

simibubi commented:
Hey, thanks a lot for the PR, and sorry to get back to it this late.
After a bit of discussion we've decided that this level of data driven-ness for fans is not in scope for base Create.
We will however happily look into some better ways to open up fan processing for add-ons, some time in the future. Either through development efforts by team members or a new PR with a smaller footprint.
Thanks again for your work. It's much appreciated

Lgmrszd commented:
@simibubi are there any internal plans to implement this? If not I might look into making this PR into less data-driven version, which would still enable other mods to add custom fan processing types

simibubi commented:
Go for it 👍
